perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -408,6 +408,94 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
 		return group;
 	}
 
+	public Group updateParentGroupId(PerunSession sess, Group group) throws InternalErrorException {
+		//return group with updated parentGroupId
+		group = getGroupsManagerImpl().updateParentGroupId(sess, group);
+
+		return group;
+	}
+
+	public void moveGroup(PerunSession sess, Group destinationGroup, Group movingGroup) throws InternalErrorException, GroupMoveNotAllowedException {
+
+		// check if moving group or destination group is null
+		if (destinationGroup == null || movingGroup == null){
+			throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " and destination group: " + destinationGroup + " cannot be null.");
+		}
+
+		// check if both groups are from same VO
+		if (destinationGroup.getVoId() != movingGroup.getVoId()) {
+			throw new GroupMoveNotAllowedException("Groups are not from same VO. Moving group: " + movingGroup + " has VO:" + movingGroup.getVoId() + " and destination group: " + destinationGroup + " has VO:" + movingGroup.getVoId() + ".");
+		}
+
+		// check if moving group is the same as destination group
+		if (destinationGroup.getId() == movingGroup.getId()) {
+			throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " cannot be the same as destination group: " + destinationGroup + ".");
+		}
+
+		// check if moving group is already in destination group
+		if (movingGroup.getParentGroupId() != null && destinationGroup.getId() == movingGroup.getParentGroupId()){
+			throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " is already under destination group: " + destinationGroup + " as subGroup.");
+		}
+
+		List<Group> movingGroupAllSubGroups = getAllSubGroups(sess, movingGroup);
+
+		// check if destination group exists as subGroup in moving group subGroups
+		if (movingGroupAllSubGroups.contains(destinationGroup)){
+			throw new GroupMoveNotAllowedException("Destination group: " + destinationGroup + " is subGroup of Moving group: " + movingGroup + ".");
+		}
+
+		// check if this operation would create cycle
+		if (checkGroupsCycle(sess, destinationGroup.getId(), movingGroup.getId())){
+			throw new GroupMoveNotAllowedException("There would be cycle created after moving group: " + movingGroup + " under destination group: " + destinationGroup + ".");
+		}
+
+		//Save old parent id for moving indirect members
+		int oldParentGroupId = movingGroup.getParentGroupId();
+		// We can move whole moving group tree under destination group
+		movingGroup.setParentGroupId(destinationGroup.getId());
+		movingGroup.setName(destinationGroup.getName()+":"+movingGroup.getShortName());
+		//we have to update group name in database
+		getGroupsManagerImpl().updateGroupName(sess, movingGroup);
```

Comment:
I believe, that in BL of normal updateGroup() or in DB ? we make sure, that on each hierarchy level, group name is unique. By moving group we could create two groups with same name on same level. Please check corresponding code to make sure this doesn't happen. Thank you.